### PR TITLE
[impeller] combine sampler and texture maps.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Simplify command encoding and reduce binding size by placing all texture/sampler data in a single map instead of 3. We don't currently (nor do we plan to) support separate textures and samplers.

The vulkan backend is particularly bad, because there are 3 map lookups to pull all of the texture and sampler data out of the bindings.

Also see [go/impeller-vulkan-cmd-recording-performance](http://goto.google.com/impeller-vulkan-cmd-recording-performance)

### CODE REVIEW

**Review by gaaclarke — COMMENTED** [review]
I don't think this is going to help, let me know if you have measured it or if I'm missing something.

**Comment by jonahwilliams** [comment]
It will help, because we're not hashing a bunch of times for the same data. Its not about the size of the map

File: impeller/renderer/command.h [hunk 1]
```diff
@@ -61,12 +61,17 @@ using BufferResource = Resource<BufferView>;
 using TextureResource = Resource<std::shared_ptr<const Texture>>;
 using SamplerResource = Resource<std::shared_ptr<const Sampler>>;
 
+/// @brief combines the texture, sampler and sampler slot information.
+struct TextureAndSampler {
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/command.h:65]: we don't need a std::variant, and its not just about the memory usage. The cost of computing hashes (std::find in the profile) is eating up CPU time.

**Review by dnfield — APPROVED** [review]

File: impeller/renderer/command.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnfield) [reply impeller/renderer/command.h:65]: Why not just go to vectors in this patch?

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/renderer/command.h:65]: baby steps :)

File: impeller/renderer/compute_command.cc [hunk 2]
```diff
@@ -61,27 +41,20 @@ bool ComputeCommand::BindResource(
...
+    return false;
+  }
+  if (!slot.HasSampler() || !slot.HasTexture()) {
     return true;
   }
 
-  bindings.samplers[slot.sampler_index] = {&metadata, sampler};
-  return true;
-}
+  bindings.sampled_images[slot.sampler_index] = TextureAndSampler{
```
Comment (gaaclarke) [inline impeller/renderer/compute_command.cc:51]: I got you, this specifically is batching the hashes at creation time for the command.  When we execute we are just looping over them anyways so it doesn't matter.  SGTM.

**Pull request merged**